src/optimizer/rule/RemoveUselessProjectRule.cpp

```diff
@@ -0,0 +1,82 @@
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "optimizer/rule/RemoveUselessProjectRule.h"
+
+#include "optimizer/OptContext.h"
+#include "optimizer/OptGroup.h"
+#include "planner/PlanNode.h"
+#include "planner/Query.h"
+
+using nebula::graph::PlanNode;
+using nebula::graph::QueryContext;
+
+namespace nebula {
+namespace opt {
+
+std::unique_ptr<OptRule> RemoveUselessProjectRule::kInstance =
+    std::unique_ptr<RemoveUselessProjectRule>(new RemoveUselessProjectRule());
+
+RemoveUselessProjectRule::RemoveUselessProjectRule() {
+    RuleSet::QueryRules().addRule(this);
+}
+
+const Pattern& RemoveUselessProjectRule::pattern() const {
+    static Pattern pattern = Pattern::create(graph::PlanNode::Kind::kProject,
+                                             {Pattern::create(graph::PlanNode::Kind::kAggregate)});
+    return pattern;
+}
+
+StatusOr<OptRule::TransformResult> RemoveUselessProjectRule::transform(
+    OptContext* octx,
+    const MatchedResult& matched) const {
+    auto qctx = octx->qctx();
+    auto* projGroupNode = matched.node;
+    DCHECK_EQ(projGroupNode->node()->kind(), PlanNode::Kind::kProject);
+    auto* oldProjNode = static_cast<const graph::Project*>(projGroupNode->node());
+
+    auto* aggGroupNode = matched.dependencies.front().node;
+    auto* oldAggNode = aggGroupNode->node();
+    DCHECK_EQ(oldAggNode->kind(), PlanNode::Kind::kAggregate);
+    auto* newAggNode = static_cast<const graph::Aggregate*>(oldAggNode)->clone(qctx);
+
+    // Delete project planNode
+    newAggNode->setOutputVar(oldProjNode->outputVar());
+    newAggNode->setInputVar(oldAggNode->inputVar());
+
+    // Fix group deps
+    auto newAggGroupNode = OptGroupNode::create(octx, newAggNode, projGroupNode->group());
+    for (auto dep : aggGroupNode->dependencies()) {
+        newAggGroupNode->dependsOn(dep);
+    }
+
+    TransformResult result;
+    result.eraseAll = true;
+    result.newGroupNodes.emplace_back(newAggGroupNode);
+    return result;
+}
+
+bool RemoveUselessProjectRule::match(OptContext* octx, const MatchedResult& matched) const {
+    UNUSED(octx);
```

Comment:
Must call `OptRule::match` firstly and then check following special logics

Reply:
Ok, I was wrong to neglect this.